### PR TITLE
fix(pi07_paligemma): gate optional prefix tokens, fix cross-att lengt…

### PR DESCRIPTION
**Pull request by akshay18iitg**
…h,
## What this does

- fix NameError

- embed_prefix now only adds response/subgoal/metadata blocks when at least one sample in the batch has real (non-pad) tokens.  Unconditional embedding injected spurious causal-block boundaries (att_masks=[1,...]) that corrupted the cumsum for every subsequent token even when all masks were False (response_drop_prob=1.0 / subgoal_drop_prob=1.0).

- Subgoal header ("Subgoal: ") and footer (", ") are now inside `if subgoal_images:`.  They were always-present real tokens that added spurious cross-attention signal and position-ID shifts in the no-subgoal case.

- num_cross_att_tokens in forward() now subtracts both the "Action: " indicator length and discrete_action_max_length, matching pi05's discrete_action_indicator_max_length logic.  Previously the indicator tokens leaked into the action expert's cross-attention at training time but were absent at inference, causing a train/inference mismatch. prefix_offsets updated to match.

- prepare_subgoal_images: track last_subgoal_img / last_mask before the empty-cameras fallback loop to avoid NameError when no subgoal keys are present in the batch and empty_cameras > 0.

Issues addressed:
- When  n_obs_step > 1 and obs_history_is_pad is True, previously only history images are zeroed out but SpaceTimeSIglip still attends to the history. This PR adds Time Attention to mask out those history images.
- When n_obs_step > 1 and obs_history_is_pad is True, the current state would also get padded. This PR fixes that
- Subgoal was using Siglip and not SpaceTimeSiglip to encode images. This PR fixes that
- The subgoal, response and metadata train time dropout where not consistent with test time dropout. This PR fixes that
- The RNG divergence error was caused due to dropout in model weights. Set dropout to 0.0 to keep the pi07_paligemma behaviour similar to pi05 under specific config

### TestPI07LowLevelPlannerObsHistoryRegression
- test_single_frame_shape_and_mask_true

  Tested: Encoder at T=1 ignores obs_history_is_pad.
  Expected: Output (B, num_video_tokens, H); None / all-True / all-False pads → identical outputs.
- test_temporal_mask_is_history_padded

  Tested: Temporal mask when only the current frame is real.
    Expected: Shape (B·N,1,T,T); past keys fully blocked; current self-only; past cannot see current; ≤1 attendable key per row; mask repeated per patch within a sample.
- test_temporal_mask_all_valid_is_causal

  Tested: Temporal mask with no padded frames.
  Expected: Standard causal lower-triangle; row i has i+1 allowed keys.
- test_temporal_mask_none_fallback_matches_all_padded

  Tested: Explicit “only current real” pad pattern vs the same pattern duplicated.
  Expected: Two explicitly built masks are exactly equal (stand-in for the forward None fallback tensor).
- test_temporal_mask_mixed_batch_per_sample_independence

  Tested: Different pad rows in one batch and a partial-history row.
  Expected: Each row’s mask equals solo build on that row; padded vs valid rows differ; partial row has the specified allow/forbid cells.

### TestPI07LowLevelPlannerStateEmbedding
- test_state_embedding_single_frame

  Tested: T=1 state block in prefix.
  Expected: Embeddings (1,1,H); state pad all present; embeddings not all zero.
- test_state_embedding_single_frame_pad_irrelevant

  Tested: T=1 with different obs_history_is_pad.
  Expected: Same state pad slice for None / True / False; still all present.
- test_state_embedding_all_history_padded
  
  Tested: T>1 with only last frame real.
  Expected: Embeddings (1,T,H); only last state slot present; last slot not all zero.
- test_state_embedding_all_pad_true_matches_history_padded

  Tested: obs_history_is_pad all True vs usual […,False] on last.
  Expected: Same state pad slice and same state embeddings (current forced present).
- test_state_embedding_all_valid

  Tested: T>1, no padding.
  Expected: All T slots present; each timestep embedding not all zero.
- test_state_embedding_none_fallback

  Tested: obs_history_is_pad None vs explicit padded-history pattern.
  Expected: Identical state pad slices.
- test_state_embedding_mixed_batch

  Tested: Three fixed per-row pad patterns.
  Expected: State pad rows [F,F,F,T], [T,T,T,T], [F,T,T,T] (presence = not pad, with last forced).
- test_forward_path_invokes_state_proj

  Tested: Multi-step state projection for the prefix path.
  Expected: Exactly one state_proj call with input shape (B,T,D).
- test_sample_actions_path_invokes_state_proj

  Tested: Single-step state projection.
  Expected: Exactly one call with shape (B,1,D).

### TestPI07LowLevelPlannerResponseEmbedding
- test_empty_response_comma_and_tokens_padded
  
  Tested: Empty response after prepare_response.
  Expected: response_masks all false; response comma and response spans in prefix pad all absent.
- test_nonempty_response_comma_and_tokens_unmasked

  Tested: Non-empty response.
  Expected: Some response_masks true; comma span all present; response span equals response_masks.
- test_mixed_batch_response_masking

  Tested: ["", real text] batch.
  Expected: Row0 comma+response absent; row1 comma present; response span equals row masks.
- test_response_masking_forward_vs_sample_actions_parity

  Tested: Stability and equivalence of preparation; empty vs real differs; missing key vs "".
  Expected: Duplicate prepares → equal tokens/masks; omit key equals explicit empty; duplicate embed_prefix pad_masks equal; empty vs real differ in comma and response spans.
### TestPI07LowLevelPlannerMetadataEmbedding
- test_all_metadata_dropped_md_comma_and_tokens_padded

Tested: All fields dropped via *_is_pad.
Expected: metadata_masks all false; metadata comma and metadata spans all absent.
- test_all_metadata_present_md_comma_unmasked

  Tested: All fields present.
  Expected: metadata_masks has some true; metadata comma all present; metadata span equals metadata_masks.
- test_metadata_is_pad_exhaustive_single_sample (8 combinations)

  Tested: Every (speed_pad, quality_pad, mistake_pad) triple vs sample_has_metadata and prefix slices.
  Expected: metadata_masks.any() iff not all three dropped; metadata comma matches metadata_masks.any(dim=1) broadcast; metadata span equals metadata_masks.
- test_mixed_batch_metadata_masking

  Tested: Three-row batch with different pad patterns.
  Expected: Comma row masks [absent, present, present] matching per-row any-metadata; each metadata row equals that row’s metadata_masks.
- test_metadata_missing_keys_matches_explicit_all_pad

  Tested: Only-state batch vs explicit all-dropped metadata batch.
  Expected: prepare_metadata outputs identical; full embed_prefix pad_masks identical.

### TestPI07LowLevelPlannerSubgoalEmbedding
- test_no_subgoal_keys_sg_block_fully_padded

  Tested: Batch with only state (no subgoal keys).
  Expected: sample_has_subgoal false; subgoal comma, header, and all subgoal vision tokens absent.
- test_subgoal_present_sg_block_unmasked

  Tested: Both subgoal cameras present and not pad-dropped.
  Expected: sample_has_subgoal true; comma+header present; each camera’s vision span equals that camera’s mask broadcast to tokens.
- test_subgoal_tensors_all_is_pad_matches_missing_keys

  Tested: Tensors + all subgoal_is_pad vs missing keys.
  Expected: prepare_subgoal_images outputs pairwise equal; embed_prefix pad_masks equal.
- test_mixed_batch_subgoal_masking

  Tested: subgoal_is_pad [True, False] with tensors for both rows.
  Expected: sample_has_subgoal == [False, True]; row0 subgoal comma/header/vision absent; row1 comma/header present; row1 full vision span present.
- test_subgoal_sample_has_or_across_cameras

  Tested: One camera all-false mask, one all-true.
  Expected: sample_has_subgoal true; subgoal comma present (OR across cameras).
- test_omit_subgoal_is_pad_defaults_all_pad

  Tested: Subgoal images present but subgoal_is_pad omitted.
  Expected: Masks all false; sample_has_subgoal false; subgoal comma span absent.
- test_subgoal_embed_prefix_parity

  Tested: Same batch prepared twice.
  Expected: embed_prefix pad_masks bitwise equal between runs.

## Checklist

- [x] I have added Google-style docstrings to important functions and ensured function parameters are typed.
- [x] My PR includes policy-related changes.
  - [x] If the above is checked: I have run the GPU pytests (pytest -m "gpu") and regression tests.

### CODE REVIEW

File: src/opentau/policies/pi07_paligemma/low_level_planner/modeling_pi07_low_level.py [hunk 1]
```diff
@@ -908,9 +908,12 @@ def prepare_subgoal_images(self, batch: dict[str, Tensor]) -> tuple[list[Tensor]
 
         # Per-sample flag: True means the subgoal was dropped or absent.
         subgoal_is_pad = batch.get(
-            "subgoal_is_pad", torch.zeros(batch["state"].shape[0], dtype=torch.bool)
+            "subgoal_is_pad", torch.ones(batch["state"].shape[0], dtype=torch.bool)
```
[the remark this answers is not in the document]
Comment (akshay18iitg) [reply src/opentau/policies/pi07_paligemma/low_level_planner/modeling_pi07_low_level.py]: This is intended

File: src/opentau/policies/pi07_paligemma/low_level_planner/modeling_pi07_low_level.py [hunk 2]
```diff
@@ -1048,17 +1051,11 @@ def __init__(self, config: PI07lowlevelPlannerConfig, discrete_action_vocab_size
...
+            multi_modal_projector=self.paligemma_with_expert.paligemma.multi_modal_projector,
             num_frames=config.n_obs_steps,
-            crop_size=config.vjepa2_crop_size,
-            num_video_tokens=config.vjepa2_num_video_tokens,
-            vlm_hidden_size=vlm_hidden_size,
-            perceiver_heads=config.vjepa2_perceiver_heads,
-            freeze_encoder=config.freeze_vision_encoder,
-            encoder_dtype=encoder_dtype,
+            spacetime_layer_stride=config.spacetime_layer_stride,
         )
```
[the remark this answers is not in the document]
Comment (shuheng-liu) [reply src/opentau/policies/pi07_paligemma/low_level_planner/modeling_pi07_low_level.py:1299]: Gradient checkpointing is important to speed up training throughput. We should keep that an option.

**Comment by shuheng-liu** [comment]
@claude fix the items raised in review

**Comment by shuheng-liu** [comment]
@claude fix all unresolved inline comments. If they are stale, reply explicitly.

**Comment by shuheng-liu** [comment]
@claude fix the pre-commit CI failures.

File: tests/policies/test_pi07_paligemma_high_level_planner.py [hunk 3]
```diff
@@ -139,6 +139,7 @@ def _verify_vlm_attention_mask(self, vlm_attention_mask, prefix_pad_masks, prefi
     # Main integration test
     # ------------------------------------------------------------------
 
+    @pytest.mark.skip(reason="run on local machine")
```
Comment (WilliamYue37) [inline tests/policies/test_pi07_paligemma_high_level_planner.py]: Please add that it is because the GPU runners don't have enough memory
